### ISSUE TITLE
Feature: Keyword filter

### ISSUE THREAD

**Issue by chbridges** [chbridges/amazon-aspect-extractor#1]
Although RAKE does a pretty good job at extracting useful keywords, the majority of the results cannot actually be used as aspects.
A possible solution is a classifier using Part-of-Speech tags as features, trained on keywords extracted from a small subset of ~100 random reviews, labelled by hand.
Will be implemented and evaluated in branch feature-keywordfilter

**Comment by chbridges** [comment]
Results for different classifiers using default parameters, a dataset of 200 hand-labelled keywords, and 5-fold Cross Validation:

Logistic Regression:
Precision: 0.7272161172161173
Recall:    0.4989646464646464
F1 score:  0.5709454796411318

Decision Tree:
Precision: 0.755988455988456
Recall:    0.4878535353535353
F1 score:  0.5789265201029907

Random Forest:
Precision: 0.7594017094017095
Recall:    0.5003535353535353
F1 score:  0.5817926195208143

Gradient Boosted Trees:
Precision: 0.7736363636363637
Recall:    0.4989646464646464
F1 score:  0.5877769289533996

Support Vector Machine:
Precision: 0.7225108225108225
Recall:    0.5500757575757576
F1 score:  0.582156446358501

Multilayer Perceptron:
Precision: 0.7319696969696972
Recall:    0.4989646464646464
F1 score:  0.575393028024607

Recall is way too low, possibly needs more tuning or better data

**Comment by chbridges** [comment]
Results when combining features "noun" and "proper noun" to a single feature:

Logistic Regression: Slightly less Precision, but general improvement.
Trees: Slightly greater Recall, but general decline.
SVM: Nope. General decline.
MLP: Nope. Just nope. Precision -10% with equal Recall.

**Comment by chbridges** [comment]
Overall improvements when using 10-fold CV:

Logistic Regression:
Precision: 0.7978571428571428
Recall:    0.5402056277056277
F1 score:  0.6253388278388278

Decision Tree:
Precision: 0.8035714285714286
Recall:    0.5402056277056277
F1 score:  0.6229578754578753

Random Forest:
Precision: 0.7849999999999999
Recall:    0.5304834054834056
F1 score:  0.6084432234432234

Gradient Boosted Trees:
Precision: 0.8214285714285715
Recall:    0.5527056277056277
F1 score:  0.6388003663003663

Support Vector Machine:
Precision: 0.7746520146520147
Recall:    0.5568722943722944
F1 score:  0.6080986119144013

Multilayer Perceptron:
Precision: 0.7197619047619048
Recall:    0.5527056277056277
F1 score:  0.6027597402597402

=> Dataset too small?

**Comment by chbridges** [comment]
As remarked by dennlinger, we should take into account that using different folds, we also evaluate the classifier on different datasets. Although only tested with default parameters, Gradient Boosting yields the best results from this collection of sklearn classifiers in each experiment. We will most likely use the XGBoost regressor and aim for larger recall using a threshold variable, i.e., we only remove keywords that are highly likely to be irrelevant for us. The code from this branch has been moved into a "prototyping" directory, however we are going to reuse the labeled dataset and parts of the original code for the new implementation of the filter.
Branch feature-keywordfilter got merged into master and deleted.

**Comment by chbridges** [comment]
Addendum: XGBoost will be replaced with the more efficient LightGBM :)

**Comment by chbridges** [comment]
The current LightGBM version seems bugged. Debug output during training can't be suppressed and CV returns an empty dictionary even though predictions clearly work. Welp. We'll just stick to XGBoost, the results were good enough.
